keiko-core/src/main/kotlin/com/netflix/spinnaker/q/metrics/QueueMetricsPublisher.kt

```diff
@@ -0,0 +1,183 @@
+import org.springframework.scheduling.annotation.Scheduled
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * - can be registered as a queue EventPublisher
+ * - publishes metrics based on queue events and regular polling of the queue state
+ */
+class QueueMetricsPublisher(
+  val registry: Registry,
+  val clock: Clock,
+  val queue: MonitorableQueue
+) : EventPublisher {
+  init {
+    PolledMeter.using(registry)
+      .withName("queue.last.poll.age")
+      .monitorValue(this, {
+        Duration
+          .between(it.lastQueuePoll, clock.instant())
+          .toMillis()
+          .toDouble()
+      })
+
+    PolledMeter.using(registry)
+      .withName("queue.last.retry.check.age")
+      .monitorValue(this, {
+        Duration
+          .between(it.lastRetryPoll, clock.instant())
+          .toMillis()
+          .toDouble()
+      })
+
+    PolledMeter.using(registry)
+      .withName("queue.depth")
+      .monitorValue(this, {
+        it.lastState.depth.toDouble()
+      })
+
+    PolledMeter.using(registry)
+      .withName("queue.unacked.depth")
+      .monitorValue(this, {
+        it.lastState.unacked.toDouble()
+      })
+
+    PolledMeter.using(registry)
+      .withName("queue.ready.depth")
+      .monitorValue(this, {
+        it.lastState.ready.toDouble()
+      })
+
+    PolledMeter.using(registry)
+      .withName("queue.orphaned.messages")
+      .monitorValue(this, {
+        it.lastState.orphaned.toDouble()
+      })
+  }
+
+  override fun publishEvent(event: QueueEvent) {
+    when (event) {
+      QueuePolled -> _lastQueuePoll.set(clock.instant())
+      is MessageProcessing -> {
+        registry.timer("queue.message.lag").record(event.lag.toMillis(), TimeUnit.MILLISECONDS)
+      }
+      is RetryPolled -> _lastRetryPoll.set(clock.instant())
+      is MessagePushed -> event.counter.increment()
+      is MessageAcknowledged -> event.counter.increment()
+      is MessageRetried -> event.counter.increment()
+      is MessageDead -> event.counter.increment()
+      is MessageDuplicate -> event.counter.increment()
+      is LockFailed -> event.counter.increment()
+      is MessageRescheduled -> event.counter.increment()
+      is MessageNotFound -> event.counter.increment()
+    }
+  }
+
+  val lastState: QueueState
+    get() = _lastState.get()
+  private val _lastState = AtomicReference<QueueState>(QueueState(0, 0, 0))
+
+  @Scheduled(fixedDelayString = "\${queue.depth.metric.frequency:30000}")
+  fun pollQueueState() {
+    _lastState.set(queue.readState())
+  }
+
+  /**
+   * Count of messages pushed to the queue.
+   */
+  private val MessagePushed.counter: Counter
+    get() = when (payload) {
+      // TODO: what to do about ApplicationAware?
+//      is ApplicationAware -> registry.counter("queue.pushed.messages", "application", (payload as ApplicationAware).application)
```

Comment:
I had a question about this, it seems ApplicationAware is an orca notion and not a keiko notion. Everything else is very orca agnostic, but I'm not sure what to do with this tag. Are we ok to drop it?

Reply:
I think this is fine to drop from keiko. Could do something like proxy the Queue class in orca and inspect the message attributes on queue.push in orca if it turns out the metric is useful.